Cython/Compiler/FusedNode.py

```diff
@@ -580,6 +600,49 @@ def _unpack_argument(self, pyx_code):
                 {{endif}}
             """)
 
+    def _fused_signature_index(self, pyx_code):
+        """
+        Generate Cython code for constructing a persistent nested dictionary index of
+        fused type specialization signatures.
+        """
+        pyx_code.put_chunk(
+            u"""
+                # CONSTRUCTING NESTED DICTIONARY INDEX OF SPECIALIZED SIGNATURES
+                if not _fused_sigindex:
+                    for sig in <dict>signatures:
+                        sigindex_node = _fused_sigindex
+                        sig_series = sig.strip('()').split('|')
+                        for sig_type in sig_series[:-1]:
+                            if sig_type not in sigindex_node:
+                                sigindex_node[sig_type] = {}
+                            sigindex_node = sigindex_node[sig_type]
+                        sigindex_node[sig_series[-1]] = sig
+            """
+        )
+    def _get_unique_identifier_base(self):
+        """
+        Try to generate a unique identifier string that will likely
+        never be repeated for different nodes, in different processes,
+        on different machines, or in different projects.
+        """
+        import time, random, sys
+        return '{seq!s}-{time_ns!s}-{id!s}-{randint!s}'.format(
+            seq = get_unique_value(),
```

Comment:
I *think* just having the `seq`entially increasing counter should be okay— provided that the counter object never gets created more than once while compiling a single module, and the call to `globals()` lower down always returns a namespace that is (at least) unique to the compiled module. The Python documentation does seem to guarantee the second assumption, but I don't know enough about Cython to know whether the first assumption is guaranteed to be true.

This code only gets run once per `cpdef` function during compilation though, so the performance impact may be inconsequential, depending on whether hashing the resulting string at `O(n)` takes up a perceptible amount of time.

Reply:
Could you use `self.node.cname`? -  that should be unique (it pretty much has to be otherwise the C code will fail to compile). I think generally it's good if you can generate the same C code each time (so that Cython-generated .c files can be included in a git repository, for example)

Reply:
`AttributeError: 'CFuncDefNode' object has no attribute 'cname'`?

Have a consistent, short, and self-commenting value guaranteed to be unique would great though. I assume it's gotta be somewhere. Where would it be stored, and would it be available at the time that `make_fused_cpdef` is called?

Reply:
+1 for some kind of `cname` prefixing. There's also the name mangling mechanism of the current scope, which adds another level of safety by generating escaped joined names like `__pyx_someprefix_4name_6length_9prepended`.

Reply:
I think `self.node.entry.cname` should work?

If that doesn't work then temporarily put `import pdb; pdb.set_trace()` just above the line, then you can use the debugger prompt and `dir` to find a suitable in `self.node`

Reply:
Take a look at what `env.mangle("prefix", "name")` returns here, and if that looks ok, drop a suitable prefix into `Naming.py` and use it to build your names like this.